patches_for_WebRTC_org/m80/0007.2-Properly-handling-async-model-for-initializing-Media.patch

```diff
@@ -0,0 +1,695 @@
++  HRESULT hr;
++
++  ComPtr<ICoreDispatcher> main_view_dispatcher;
++  ComPtr<IAsyncAction> async_action_ui_dispatcher;
++  ComPtr<IAsyncAction> async_action_media_capture;
++  ComPtr<IMediaCapture> media_capture;
++  AgileRef media_capture_agile;
++  AgileRef async_action_media_capture_agile;
++  Event event_wait_for_media_capture_async_action;
++  Event event_wait_for_media_capture_async_action_completion;
++  boolean has_thread_access;
++
++  // Acquires the main view dispacther (UI thread).
++  THR(GetMainViewDispatcher(&main_view_dispatcher));
++
++  // We'll be dispatching and waiting for code from the UI thread.
++  // Let's make sure this is not the UI thread.
++  THR(main_view_dispatcher->get_HasThreadAccess(&has_thread_access));
++  THR(has_thread_access ? E_FAIL : S_OK);
++
++  // The media capture device has to be initialized in the UI thread.
++  // The following event is used for letting this thread know that
++  // async_action_media_capture has been created.
++  event_wait_for_media_capture_async_action =
++      Event(::CreateEventEx(nullptr, nullptr, CREATE_EVENT_MANUAL_RESET,
++                            WRITE_OWNER | EVENT_ALL_ACCESS));
++  THR(event_wait_for_media_capture_async_action.IsValid()
++          ? S_OK
++          : HRESULT_FROM_WIN32(GetLastError()));
++
++  THR(ActivateInstance(
++      HStringReference(RuntimeClass_Windows_Media_Capture_MediaCapture).Get(),
++      media_capture.ReleaseAndGetAddressOf()));
++
++  THR(media_capture.AsAgile(&media_capture_agile));
++
++  // Dispatches the media capture initialization to the main view
++  // dispatcher.
++  THR(main_view_dispatcher->RunAsync(
++      CoreDispatcherPriority::CoreDispatcherPriority_Normal,
++      Callback<Implements<RuntimeClassFlags<Delegate>, IDispatchedHandler,
++                          FtmBase>>([&event_wait_for_media_capture_async_action,
++                                     &async_action_media_capture_agile,
++                                     &media_capture_agile,
++                                     &device_id]() -> HRESULT {
++        HRESULT hr;
++
++        ComPtr<IMediaCapture> media_capture;
++        ComPtr<IAsyncAction> media_capture_async;
++        ComPtr<IMediaCaptureInitializationSettings> init_settings;
++        ComPtr<IMediaCaptureInitializationSettings5> init_settings5;
++
++        THR(media_capture_agile.As(&media_capture));
++
++        // Creates the settings used to select which capture device will be
++        // used.
++        THR(ActivateInstance(
++            HStringReference(
++                RuntimeClass_Windows_Media_Capture_MediaCaptureInitializationSettings)
++                .Get(),
++            &init_settings));
++        THR(init_settings->put_StreamingCaptureMode(
++            StreamingCaptureMode::StreamingCaptureMode_AudioAndVideo));
++        THR(init_settings.As<IMediaCaptureInitializationSettings5>(
++            &init_settings5));
++        THR(init_settings5->put_MemoryPreference(
```

Comment:
I assume this will go away with the no-copy optimization, since we want the frame to stay in VRAM until it's encoded, no?